rfcs/0053-reducing-daemon-memory-usage.md

```diff
@@ -0,0 +1,137 @@
+
+## Memory Analysis
+[memory-analysis]: #memory-analysis
+
+<!-- TODO: better name for this script so that it's not so custom-tailored to this conversation -->
+In order to accurately estimate the memory usage of the daemon in the fully saturated max-cost transaction environment, we have written a program `src/app/ram_fix_math/ram_fix_math.exe`. This program counts the size of GC allocations on various data structures used by the daemon, and does so by carefully ensuring every value is a unique allocation and that there are no shared references within data structures. We do this by transporting values back and forth via bin_prot, simulating the same behavior the daemon will have when it reads and deserializes data from the network. We then use these measurements to estimate the expected worst-case memory footprint of larger data structures in the system, such as the mempools and the frontier. Expectations around shared references across these larger data structures are directly subtracted from the estimates.
+
+Below is the output of the script when all of the parameters are tuned to what we have planned for the Berkeley release.
+
+```
+baseline = 3.064569GB
+scan_states = 15.116044GB
+ledger_masks = 2.292873GB
+staged_ledger_diffs = 32.345724GB
+snark_pool = 4.453962GB
+transaction_pool = 0.871383GB
+TOTAL: 58.144555GB
+```
+
+In this output, the baseline is made up of static measurements taken from the daemon, which represents the overhead of running a daemon regardless of throughput or transaction cost. We have then estimated the expected worst-case memory usage of the scan states, ledger masks, and staged ledger diffs, which are the overwhelming majority of data allocated for the frontier. And finally, we estimate the memory footprint of the snark pool and transaction pool.
+
+Now, if we adjust the estimates for proposed optimizations to store proofs and verification keys to disk (by subtracting their sizes from the memory footprint and replacing them with cache references), we get the following output.
+
+```
+baseline = 3.064569GB
+scan_states = 3.658435GB
+ledger_masks = 0.562126GB
+staged_ledger_diffs = 9.202166GB
+snark_pool = 0.014414GB
+transaction_pool = 0.247903GB
+TOTAL: 16.749612GB
+```
+
+As we can see, this brings the estimation down much closer to the current `16GB` hardware requirement. From here, we can look into additional optimizations to bring it down even further to fit within the current hardware requirements. Such optimizations could include:
```

Comment:
Since there is already some discussion of raising the requirements slightly, I propose to defer all of the following extra optimizations until later.